Add "Advanced `mailto:` with Symfony Mailer" post (#50)
knpuniversity/blog: advanced-mailto-with-symfony-mailer.md

# Ditch mailto Links: Pre-Fill & even Attach Files with Symfony Mailer (.eml Magic!)

Congratulations on your new gig working on the site for "Prehistoric Pens": the leader

> in building dinosaur enclosures!

Reviewer: ```suggestion
in dinosaur enclosures!
```

> Every morning (after getting a soda) our sales person Dennis reviews customer requests

Reviewer: ```suggestion
Each morning (after getting a soda) our sales person Dennis reviews customer requests
```

asking for a quote on a specific pen. To be personal (and see it in his sent messages),
Dennis copies the customer email address, composes an email in his own email client and

> types out a personal message. He also finds and attaches the specs PDF of the pen they want.

Reviewer: ```suggestion
types out a personal message. He also finds and attaches the "specs" PDF of the pen the customer wants.
```

        ->attach($pdf->read(), $pdf->path()->name(), $pdf->mimeType())
    ;

    $response = new Response($email->toString());
    $contentDisposition = $response->headers->makeDisposition(ResponseHeaderBag::DISPOSITION_ATTACHMENT, $quote->getId().'.eml');
    
    $response->headers->set('Content-Type', 'message/rfc822');
    $response->headers->set('Content-Disposition', $contentDisposition);
    
    return $response;
}
```

Let's unpack this:
1. We create an email like normal... but using a `DraftEmail` object. This
   is a special kind of email that cannot be sent with Mailer directly.
2. Add the Quote PDF attachment also like normal.
3. We then send an *attachment* response. The content is the *stringified*
   email. The content type is `message/rfc822` which is the MIME type for an email
   (`.eml` file).

Now, on our quote page, instead of a `mailto:` link, we can link to this controller:

```twig
<a href="{{ path('quote_followup_email', {id: quote.id}) }}">
    Send Follow-up
</a>
```

When Dennis clicks this link, his browser will download a `.eml` file.
Opening this will open his email client with the `to` email pre-filled and the PDF attached.

> He can adjust the message, maybe add his signature, and send it off!

Reviewer: All he needs to do is add the message (written with the famous Dennis Nedry charm), hit send and finish that tasty soda. Another successful day selling dino pens Dennis!

> He can adjust the message, maybe add his signature, and send it off!

Author: ```suggestion
All he needs to do is add the message (written with the famous Dennis Nedry charm),
hit send and finish that tasty soda. Another successful day selling dino pens Dennis!
```

See the Symfony Documentation for more information about
[Draft Emails](https://symfony.com/doc/current/mailer.html#draft-emails).

> Check out our full and free Symfony

Reviewer: ```suggestion
And check out our full and free Symfony
```

[Mailer and Webhook with Mailtrap](https://symfonycasts.com/screencast/mailtrap)

> course for more email fun!

Reviewer: ```suggestion
course for more email goodies!
```

> # Advanced `mailto:` with Symfony Mailer

Reviewer: ```suggestion
Ditch mailto Links: Pre-Fill & even Attach Files with Symfony Mailer (.eml Magic!)
```

Consider the following scenario: you manage the website for a company that
builds and sells dinosaur enclosures: "Prehistoric Pens".

There's a section for the sales team where they can review pending
quotes. They want a new feature: when they view a quote, they want to

> quickly send a follow-up email to the customer.

Author: I don't want to get into the weeds with a fictitious app but where would the prefilled customer email come from in your scenario?

The quote email system is something I've actually built (not for dino pens). I get that means is makes perfect sense to me but perhaps not others.
